cassiopeia/solver/SpectralNeighborJoiningSolver.py

```diff
@@ -0,0 +1,318 @@
+from cassiopeia.solver import (
+    dissimilarity_functions,
+    solver_utilities,
+)
+
+
+class SpectralNeighborJoiningSolver(DistanceSolver):
+    """
+    Spectral Neighbor-Joining class for Cassiopeia.
+
+    Implements a variation on the Spectral Neighbor-Joining
+    algorithm described by Jaffe et al. in their 2020 paper. This class
+    implements the abstract class DistanceSolver. This class inherits
+    the 'solve' method from DistanceSolver with the 'find_cherry' method
+    implemented using a spectral method as in Jaffe et al.  (2020).
+
+    Args:
+        similarity_function: A function by which to compute the similarity map,
+            corresponding to the affinity R(i, j) in Jaffe et al.
+            (2020). Note that this similarity score should be multiplicative,
+            i.e. R(i, j) = R(i, k) * R(k, j) for k on the path from i to j on
+            the lineage tree. By default we will use exp(-d(i,j)) where d is the
+            metric given by weighted_hamming_distance from
+            dissimilarity_functions without weights given.
+
+        add_root: Whether or not to add an implicit root to the tree, i.e. a
+            root with unmutated characters.
+        prior_transformation: Function to use when transforming
+            priors into weights
+    """
+
+    def __init__(
+        self,
+        similarity_function: Optional[
+            Callable[
+                [np.ndarray, np.ndarray, int, Dict[int, Dict[int, float]]],
+                float,
+            ]
+        ] = dissimilarity_functions.exponential_negative_hamming_distance,
+        # type: ignore
+        add_root: bool = False,
+        prior_transformation: str = "negative_log",
+    ):
+        super().__init__(
+            dissimilarity_function=similarity_function,
+            add_root=add_root,
+            prior_transformation=prior_transformation,
+        )  # type: ignore
+
+    def get_dissimilarity_map(
+        self, cassiopeia_tree: CassiopeiaTree, layer: Optional[str] = None
+    ) -> pd.DataFrame:
+        """Outputs the first lambda matrix, where every subset is an
+        individual node.
+
+        Args:
+            cassiopeia_tree: the CassiopeiaTree object passed into solve()
+            layer: Layer storing the character matrix for solving. If None, the
+                default character matrix is used in the CassiopeiaTree.
+
+        Returns:
+            DataFrame object of the lambda matrix, but similar in structure to
+                DistanceSolver's dissimilarity_map.
+        """
+
+        # get dissimilarity map and save it as private instance variable
+        self._similarity_map = super().get_dissimilarity_map(
+            cassiopeia_tree, layer
+        )
+
+        # prepare for first (pairwise) lambda matrix
+        N = self._similarity_map.shape[0]
+        node_names: np.ndarray = self._similarity_map.index.values
+        self.lambda_indices = [[i] for i in range(N)]
```

Comment:
Neither `_similarity_map` nor `lambda_indices` are instantiated in the __init__ method. You should instantiate them as `None` in the constructor.

Also, should we make `lambda_indices` a private variable as well? Theoretically, this should be removed after each run of the SNJ Solver and a user shouldn't be able to access it.